elasticdl/pkg/common/vector.go

```diff
@@ -0,0 +1,235 @@
+	case float64:
+		binary.LittleEndian.PutUint64(buffer[idx*size:idx*size+size], math.Float64bits(val.(float64)))
+	}
+}
+
+// ConstantInit return a constant iniitializer
+func ConstantInit(val interface{}) InitialzeFunc {
+	return func(buffer []byte) error {
+		size := int(reflect.ValueOf(val).Type().Size())
+		length := len(buffer) / size
+		for idx := 0; idx < length; idx++ {
+			byteSet(buffer, idx, size, val)
+		}
+		return nil
+	}
+}
+
+// CopyInit return a copy initializer
+func CopyInit(slice interface{}) InitialzeFunc {
+	return func(b []byte) error {
+		buffer := bytes.NewBuffer(b)
+		buffer.Reset()
+		return binary.Write(buffer, binary.LittleEndian, slice)
+	}
+}
+
+// NewEmptyVector create an zero-initialized vector
+func NewEmptyVector(length int, dtype DataType) *Vector {
+	bytelen := int(dtype.Size) * length
+	var vec = Vector{
+		Data:   make([]byte, bytelen, bytelen),
+		Dtype:  dtype,
+		Length: length,
+	}
+	return &vec
+}
+
+// NewInitializerVector create an initialized vector
+func NewInitializerVector(length int, dtype DataType, initializer InitialzeFunc) *Vector {
+	bytelen := int(dtype.Size) * length
+	data := make([]byte, bytelen, bytelen)
+	initializer(data)
+	var vec = Vector{
+		Data:   data,
+		Dtype:  dtype,
+		Length: length,
+	}
+	return &vec
+}
+
+// NewVectorFrom create a vector using a slice to initialize
+func NewVectorFrom(slice interface{}) *Vector {
+	v := reflect.ValueOf(slice)
+	dtype := TypeToDataType[v.Type()]
+	length := v.Len()
+	initializer := CopyInit(slice)
+	return NewInitializerVector(length, dtype, initializer)
+}
+
+// At get the element at an index
+func (v *Vector) At(idx int) float64 {
+	size := v.Dtype.Size
+	switch v.Dtype.Flag {
+	case Int8:
+		return float64(int8(v.Data[idx]))
+	case Int16:
+		return float64(int16(binary.LittleEndian.Uint16(v.Data[idx*size : idx*size+size])))
+	case Int32:
+		return float64(int32(binary.LittleEndian.Uint32(v.Data[idx*size : idx*size+size])))
+	case Int64:
+		return float64(int64(binary.LittleEndian.Uint64(v.Data[idx*size : idx*size+size])))
+	case Float32:
+		return float64(math.Float32frombits(binary.LittleEndian.Uint32(v.Data[idx*size : idx*size+size])))
+	case Float64:
+		return math.Float64frombits(binary.LittleEndian.Uint64(v.Data[idx*size : idx*size+size]))
+	default:
+		return 0
+	}
+}
+
+// Set set the value to an index
+func (v *Vector) Set(idx int, val interface{}) {
+	size := v.Dtype.Size
+	byteSet(v.Data, idx, size, val)
+}
+
+// InplaceSlice gives a Slice interface to the Vector data
+func (v *Vector) InplaceSlice() interface{} {
+	sliceHeader := reflect.SliceHeader{
+		Data: uintptr(unsafe.Pointer(&v.Data[0])),
+		Cap:  int(v.Length),
+		Len:  int(v.Length),
+	}
+	val := reflect.NewAt(v.Dtype.Type, unsafe.Pointer(&sliceHeader)).Elem()
+	return val.Interface()
+}
+
+// MakeSlice gives a slice copy of the Vector data
+func (v *Vector) MakeSlice() interface{} {
```

Comment:
Maybe we could merge `InplaceSlice` and `MakeSlice` into one function `ToSlice`

```go
func (b *Buffer) ToSlice(isCopy bool) interface {} {
   ......
}
```

Reply:
when writing codes, Maybe `vec.ToSlice(true)` is not so obvious which of `inplace` or `copy` is true?